flake8_type_checking/checker.py

```diff
@@ -360,28 +361,147 @@ def import_type(self) -> ImportTypeValue:
+        self.symbols: dict[str, list[Symbol]] = defaultdict(list)
+
+        #: The outer scope, this will be `None` for the global scope
+        # We use this to traverse up to the outer scopes when looking up
+        # symbols
+        self.parent = parent
+
+        #: For function scopes/class scopes whether it is just for the head or also the body
+        self.is_head = is_head
+
+        #: The name of the class if this is a scope created by a class definition
+        # classes are not real scopes, i.e. they don't propagate symbols
+        # to inner-scopes, so we need to treat them differently in lookup
+        # the class name itself is also special, since it's available in methods
+        # but not in the class body itself, so we record it, so we can special-case
+        # it in symbol lookups
+        self.class_name = node.name if isinstance(node, ast.ClassDef) else None
+
+    def lookup(self, symbol_name: str, use: HasPosition | None = None, runtime_only: bool = True) -> Symbol | None:
+        """
+        Simulate a symbol lookup.
+
+        If a symbol is redefined multiple times in the same block we don't try
+        to return the symbol closest to the use-site, we just return the first
+        one we find, since we don't really care what symbol we find currently.
+        """
+        for symbol in self.symbols.get(symbol_name, ()):
+            if runtime_only and not symbol.available_at_runtime(use):
+                continue
+
+            # we just return the first symbol we find
+            return symbol
+
+        parent = self.parent
+        if runtime_only:
+            # if the symbol matches our class name we return at this point
+            # technically if the class definition is a redefinition of the
+            # the same symbol_name it could still exist at runtime, but it's
+            # probably an actual mistake at that point and an annotation should
+            # be quoted to ensure the correct type is assigned
+            if symbol_name == self.class_name:
+                return None
+
+            if parent is not None:
+                # because of our virtual scope for function definition headers
+                # we also need to check the parent scope for a class name
+                if self.is_head and symbol_name == parent.class_name:
+                    return None
+
+                # if the next scope up is a head scope we don't want to skip it
+                if parent.is_head:
+                    return parent.lookup(symbol_name, use, runtime_only)
+
+            # skip class scopes when looking up symbols in parent scopes
+            # they're only available inside the class scope itself
+            while parent is not None and parent.class_name is not None:
+                parent = parent.parent
+
+        # we're done looking up and didn't find anything
+        if parent is None:
+            return None
+
+        return parent.lookup(symbol_name, use, runtime_only)
```

Comment:
Pretty much, although I'd consider scopes separately from the AST, since the scopes will be fully populated once we do these lookups, we don't do any lookups while we walk the AST. But the scopes will have at least partially the same tree structure as the AST, since they're created by specific AST nodes.

For runtime lookups we technically only care about the declarations above us but for type checking lookups the definition can occur anywhere inside the relevant scopes.

Reply:
I just realized the `use` should not always propagate to outer scopes. For real scopes like function scopes we don't care where the function is defined in order to access globals within the body. So that will need to be fixed. (It still needs to propagate for the virtual head scope though, since that's technically part of the outer scope, it just also contains symbols that are not allowed to leak into the real scope)

Reply:
Function bodies are late binding, they look up the variable when they're not called, not when they're defined. So it's perfectly legal to defined a global variable after a function that uses it. There are of course scenarios where it would still raise a `NameError` (i.e. if you call the function before you declare the symbol) but it's difficult to statically analyze this, so it's best to assume a symbol exists at runtime as long as it's found in an outer scope at all, pyflakes does the same thing.

```python
>>> def foo():
...     return x
... 
>>> foo()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in foo
NameError: name 'x' is not defined
>>> x = 5
>>> foo()
5
>>> 
```
I did fix that however in the meantime, `use` gets set to `None` when doing a runtime lookup and popping up a real scope.